exodus_lambda/functions/verify_source.py

```diff
@@ -0,0 +1,64 @@
+import os
+import subprocess
+
+import boto3
+
+
+def lambda_handler(event, context):
+    """Verifies the commit delivered to the pipeline by GitHub has a
+    tag signed by a trusted entity.
+
+    Receives the commit ID from the pipeline via UserParameters.
+
+    This function must be manually deployed prior to pipeline creation.
+    """
+    # pylint: disable=unused-argument
+
+    job_id = event["CodePipeline.job"]["id"]
+    job_data = event["CodePipeline.job"]["data"]
+    git_commit = job_data["actionConfiguration"]["configuration"][
+        "UserParameters"
+    ]
+
+    bucket_name = "exodus-pipeline-artifacts"
+    whitelist_name = ".whitelist.gpg"
+
+    s3_client = boto3.client("s3")
+    pipeline_client = boto3.client("codepipeline")
+
+    try:
+        # 1) Download and import whitelisted GPG keys
+        s3_client.download_file(bucket_name, whitelist_name)
+        subprocess.run(
+            ["gpg", "--import", whitelist_name], check=True, text=True
+        )
+
+        # 2) Clone exodus-lambda and checkout deploy branch
+        repo_url = "https://github.com/release-engineering/exodus-lambda.git"
```

Comment:
Do we really need to clone it again... don't we already have a SourceArtifact with the repo?  However I admittedly don't know whether that SourceArtifact zip file includes the .git directory or only the working tree.  If it doesn't include .git then indeed we'd have to clone.

One of the downsides of cloning it again is that you make it harder to test locally and harder to point it at some other repo (e.g. remembering Caleigh's PR where she was hacking it to point at her fork).

Reply:
Yeah, like the previous attempt at GPG verification in the buildspec, we have to clone the repo to access the git metadata, as it isn't included in the source zip sent from GitHub.

Reply:
In that case, if it's not much trouble do you think you could add the URL and branch into the pipeline and pass them in as UserParameters as well?  Or: is it possible to extract them from the Source action's output variables?  According to https://docs.aws.amazon.com/codepipeline/latest/userguide/action-reference-GitHub.html#action-reference-GitHub-variables , it will export RepositoryName and BranchName, so it seems like it should be possible to reuse them rather than hardcoding here.  Or maybe they are even in job_params already like the commit_id is?

It would be nice to avoid someone having to hack this to point at their own repo if they want to test from their own fork.

Reply:
Agreed.
The branch and repo names from the source stage variables are usable. Though, one would still need to change owner, repo and branch values in source stage to point the pipeline elsewhere. So, I'll just create owner, repo and branch template parameters and supply them to source and verify stages.